packages/opencensus-instrumentation-grpc/src/grpc.ts

```diff
@@ -0,0 +1,478 @@
+      };
+      return plugin.tracer.wrap(wrappedFn);
+    }
+
+    return (span: types.Span) => {
+      if (!span) {
+        return original.apply(self, args);
+      }
+
+      // if unary or clientStream
+      if (!original.responseStream) {
+        const callbackFuncIndex = findIndex(args, (arg) => {
+          return typeof arg === 'function';
+        });
+        if (callbackFuncIndex !== -1) {
+          args[callbackFuncIndex] =
+              patchedCallback(span, args[callbackFuncIndex]);
+        }
+      }
+
+      const metadata = this.getMetadata(original, args, span);
+
+      const setter: types.HeaderSetter = {
+        setHeader(name: string, value: string) {
+          metadata.set(name, value);
+        }
+      };
+
+      const propagation = plugin.tracer.propagation;
+      if (propagation) {
+        propagation.inject(setter, span.spanContext);
+      }
+
+      span.addAttribute(GrpcPlugin.ATTRIBUTE_GRPC_METHOD, original.path);
+      span.addAttribute(GrpcPlugin.ATTRIBUTE_GRPC_KIND, span.type);
+
+      const call = original.apply(self, args);
+
+      plugin.tracer.wrapEmitter(call);
+
+      // if server stream or bidi
+      if (original.responseStream) {
+        let spanEnded = false;
+        call.on('error', (err: grpcModule.ServiceError) => {
+          // span.status = plugin.traceStatus(err.code);
+          span.addAttribute(GrpcPlugin.ATTRIBUTE_GRPC_ERROR_NAME, err.name);
+          span.addAttribute(
+              GrpcPlugin.ATTRIBUTE_GRPC_ERROR_MESSAGE, err.message);
+          if (!spanEnded) {
+            span.end();
+            spanEnded = true;
+          }
+        });
+
+        call.on('status', (status: Status) => {
+          span.status = plugin.traceStatus(status.code);
+          span.addAttribute(
+              GrpcPlugin.ATTRIBUTE_GRPC_STATUS_CODE, status.code.toString());
+
+          if (!spanEnded) {
+            span.end();
+            spanEnded = true;
+          }
+        });
+      }
+      return call;
+    };
+  }
+
+  /**
+   * Gets a metadata form args, creates a new one if not found.
+   *
+   *  Code snippet inspired by:
+   *  https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/src/plugins/plugin-grpc.ts#L96)
+   */
+  // tslint:disable-next-line:no-any
+  private getMetadata(original: any, args: any[], span: types.Span):
+      grpcModule.Metadata {
+    let metadata: grpcModule.Metadata;
+
+    // This finds an instance of Metadata among the arguments.
+    // A possible issue that could occur is if the 'options' parameter from
+    // the user contains an '_internal_repr' as well as a 'getMap' function,
+    // but this is an extremely rare case.
+    // tslint:disable-next-line:no-any
+    let metadataIndex = findIndex(args, (arg: any) => {
+      return arg && typeof arg === 'object' && arg._internal_repr &&
+          typeof arg.getMap === 'function';
+    });
+    if (metadataIndex === -1) {
+      metadata = new grpcModule.Metadata();
```

Comment:
We should avoid using `grpcModule` for constructing an actual `Metadata` object, if possible.

Reply:
PR #52